[pipelining] allow multiple backward grads

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140981

fixes https://github.com/pytorch/pytorch/issues/139404. The input grads get saved in a new `self.bwd_cache` container and get popped off after they are used in `backward_one_chunk`

`python test/distributed/pipelining/test_schedule_multiproc.py -k test_pipeline_schedule_runtime_custom_sched`

cc @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
[pipelining] allow multiple backward grads

[ghstack-poisoned]

wconstab commented on test/distributed/pipelining/schedule_registry.py:
```diff
@@ -172,3 +172,50 @@ def __init__(
+                _Action(2, F, 1),
+                None,
+                None,
+                _Action(2, B, 0),
```
i'm probably just missing something but isn't this schedule ordered in the usual order,  i.e. not reordered B?

the usual order would execute B's in increasing microbatch_id sequence right?

H-Huang replied:
Yeah good point, I realized the test is not good because the error would only surface if we use send for mb A after we did a backward for mb B, which the test is not catching. So i updated it to reorder the send operations in the new schedule

wconstab reviewed (approved):
code change LGTM. just want to make sure the test case tests what it's supposed to

H-Huang committed:
Update on "[pipelining] allow multiple backward grads"


fixes https://github.com/pytorch/pytorch/issues/139404. The input grads get saved in a new `self.bwd_cache` container and get popped off after they are used in `backward_one_chunk`

`python test/distributed/pipelining/test_schedule_multiproc.py -k test_pipeline_schedule_runtime_custom_sched`

cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge

H-Huang committed:
Update on "[pipelining] allow multiple backward grads"


fixes https://github.com/pytorch/pytorch/issues/139404. The input grads get saved in a new `self.bwd_cache` container and get popped off after they are used in `backward_one_chunk`

`python test/distributed/pipelining/test_schedule_multiproc.py -k test_pipeline_schedule_runtime_custom_sched`

cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge
